### ISSUE TITLE
Add a crypt() function supporting SHA256 and SHA512

### ISSUE THREAD

**Issue by pgaxatte** [hashicorp/terraform#26059]
### Current Terraform Version
```
Terraform v0.13.1
```

### Use-cases
bcrypt() allows the generation of salted hashes respecting the glibc format but only supports blowfish which is generally not supported in glibc (cf. man 3 crypt).

It would be nice to have a crypt() function that support at least SHA-256 and SHA-512 to be able to use it with products that don't support blowfish (HAProxy for instance, see [their documentation](http://cbonte.github.io/haproxy-dconv/2.2/configuration.html#3.4-user))

**Comment by crw** [comment]
Hi @cybernetic-mountaineer, thanks for your comment. There is no plan to add additional cryptographic functions directly to Terraform as built-in functions. That said, a cryptographic function library is one of the main use cases of implementing providers for functions (aka user-provided functions), per https://github.com/hashicorp/terraform/issues/27696. I am sorry I don't have anything more concrete to offer at this time, providers for functions is going to be the path forward for these and many other issues around implementing specific functions. Thanks again!

**Comment by cybernetic-mountaineer** [comment]
Heard, thanks @crw

**Comment by cybernetic-mountaineer** [comment]
So, great news :) There's a provider that supports this! 
https://github.com/loafoe/terraform-provider-htpasswd
https://registry.terraform.io/providers/loafoe/htpasswd/

Here's the relevant issue: https://github.com/loafoe/terraform-provider-htpasswd/issues/5

**Comment by crw** [comment]
Thank you for your continued interest in this issue. 

[Terraform version 1.8](https://github.com/hashicorp/terraform/blob/v1.8.0-beta1/CHANGELOG.md) launches with support of provider-defined functions. It is now possible to implement your own functions! We would love to see this implemented as a provider-defined function. 

Please see the [provider-defined functions documentation](https://developer.hashicorp.com/terraform/plugin/framework/functions) to learn how to implement functions in your providers. If you are new to provider development, learn how to [create a new provider](https://developer.hashicorp.com/terraform/plugin/framework) with the Terraform Plugin Framework. If you have any questions, please visit the [Terraform Plugin Development category in our official forum](https://discuss.hashicorp.com/c/terraform-providers/tf-plugin-sdk/43). 

We hope this feature unblocks future function development and provides more flexibility for the Terraform community. Thank you for your continued support of Terraform!
